src/main/resources/category/java/sql.xml

```diff
@@ -0,0 +1,165 @@
+    (: good: builder.in(cust.get("postalCode")).value(subquery) :)
+    (not(exists(ancestor::PrimaryExpression/PrimarySuffix/Arguments//Expression[pmd-java:typeIs('javax.persistence.criteria.Subquery')])))]
+    ]]>
+                    </value>
+                </property>
+            </properties>
+            <example>
+                <![CDATA[
+        private void bad(CriteriaQuery cq, Root<Customer> cust) {
+            cq.select(cust).where(cust.get("postalCode").in(codes)); // bad
+        }
+        private void good(CriteriaBuilder builder, CriteriaQuery cq, Root<Customer> cust, SubQuery<PostalCode> subquery) {
+            cq.select(cust).where(builder.in(cust.get("postalCode")).value(subquery));
+        }
+                ]]>
+            </example>
+        </rule>
+-->
+    <rule name="AvoidFetchingWholeList"
+          message="Avoid fetching the whole list of entities for the query, while only one entity is used"
+          class="net.sourceforge.pmd.lang.rule.xpath.XPathRule"
+          language="java"
+          externalInfoUrl="${doc_root}/JavaDataAccessPerformance.md#ida-trr04">
+        <description>
+            Problem: Unused rows are fetched and transported, and unused jdbc buffer is allocated.&#13;
+            Solution: Use query.getSingleResult() in stead of query.getResultList().
+            (jpinpoint-rules)</description>
+        <priority>2</priority>
+        <properties>
+            <property name="tags" value="jpinpoint-rule,performance" type="String" description="classification"/>
+            <property name="xpath">
+                <value>
+                    <![CDATA[
+(: if getResultList is used to fetch entities and entities(0) is returned :)
+//ReturnStatement//MethodCall[pmd-java:matchesSig('java.util.List#get(int)')][./ArgumentList/NumericLiteral[@ValueAsInt=0]]/VariableAccess[
+    @Name = ancestor::MethodDeclaration//VariableDeclarator[
+        MethodCall[pmd-java:matchesSig('javax.persistence.TypedQuery#getResultList()') or pmd-java:matchesSig('jakarta.persistence.TypedQuery#getResultList()')]
+    ]/VariableId/@Name
+    and not(
+        (: not when also another elem of entities is used in the method :)
+        @Name = ancestor::MethodDeclaration//MethodCall[pmd-java:matchesSig('java.util.List#get(int)')][./ArgumentList/NumericLiteral[@ValueAsInt!=0]]/VariableAccess/@Name
+        or
+        (: not when also the whole list of entities is returned :)
+        @Name = ancestor::MethodDeclaration//ReturnStatement//(ConditionalExpression|.)/VariableAccess[not(parent::MethodCall)]/@Name
+    )
+]
+]]>
+                </value>
+            </property>
+        </properties>
+        <example>
+            <![CDATA[
+    public SomeEntity findByQueryBad(TypedQuery<Object> query) {
+        final List<SomeEntity> entities = query.getResultList();
+        return !entities.isEmpty() ? entities.get(0) : null;
```

Comment:
Note there are other ways to get first element: 

- java 21 has `getFirst()`, which intellij also suggests as improved code. 
- via iterator: `entities.iterator().next()`
- via stream: `entities.stream().findFirst().get()` or other stream constructs with `findAny()`

Reply:
Maybe getFirst() match can be added ... others via issue for future?

Reply:
okay, good to know, new functionality, for improvement request

Reply:
Added #377